cosign/private/sign.bzl

```diff
@@ -0,0 +1,77 @@
+"Implementation details for sign rule"
+
+_DOC = """Sign an oci_image using cosign binary at a remote registry.
+
+It sings the image by its digest determined beforehand.
+
+```starlark
+oci_image(
+    name = "image"
+)
+
+cosign_sign(
+    name = "sign",
+    image = ":image",
+    repository = "index.docker.io/org/image"
+)
+```
+
+`repository` attribute can be overridden using the `--repository` option.
+
+```starlark
+oci_image(
+    name = "image"
+)
+
+cosign_sign(
+    name = "sign",
+    image = ":image",
+    repository = "index.docker.io/org/image"
+)
+```
+
+run `bazel run :sign -- --repository=index.docker.io/org/test`
+"""
+
+_attrs = {
+    "image": attr.label(allow_single_file = True, doc = "Label to an oci_image"),
+    "repository": attr.string(mandatory = True, doc = "Repository URL where the image will be signed at. eg: index.docker.io/<user>/image. digests and tags are disallowed."),
+    "_sign_sh_tpl": attr.label(default = "sign.sh.tpl", allow_single_file = True),
+}
+
+def _cosign_sign_impl(ctx):
+    cosign = ctx.toolchains["@contrib_rules_oci//cosign:toolchain_type"]
+    yq = ctx.toolchains["@aspect_bazel_lib//lib:yq_toolchain_type"]
+
+    if ctx.attr.repository.find(":") != -1 or ctx.attr.repository.find("@") != -1:
+        fail("repository attribute should not contain digest or tag.")
+
+    executable = ctx.actions.declare_file("cosign_%s" % ctx.label.name)
+    ctx.actions.expand_template(
+        template = ctx.file._sign_sh_tpl,
+        output = executable,
+        is_executable = True,
+        substitutions = {
+            "{{cosign_path}}": cosign.cosign_info.cosign_path,
+            "{{yq_path}}": yq.yqinfo.bin.path,
```

Comment:
related to other comment, if you pass the binary File to the toolchain info then you choose between path and short_path here...

here this should be `yq.yqinfo.bin.short_path` so you're not depending on legacy external runfiles